### PR TITLE
ARROW-17439: [R] Change behavior of pull to compute instead of collect

### PR DESCRIPTION
**Pull request by amoeba**
I could use some help with a couple of things:

- [x] `devtools::check` warns about `pull.ArrowTabluar` being undocumented. I `@export`ed it to stay consistent with other ArrowTabular generics defined in `arrow-tabular.R` and don't understand why checking doesn't warn on all of these. Does this one just not need exporting?
    <details>
    <summary>Relevant devtools::check() output</summary>

    ```
    ❯ checking for missing documentation entries ... WARNING
      Undocumented code objects:
        ‘pull.ArrowTabular’
      All user-level objects in a package should have documentation entries.
      See chapter ‘Writing R documentation files’ in the ‘Writing R
      Extensions’ manual.
    ```
    </details>
- [x] I found an inconsistency with `dplyr::pull` and Tables: Pulling an ungrouped Table produces a ChunkedArray whereas pulling a grouped Table produces a Table. This makes a subsequent call to `as.vector` produce an error of `Error in as.vector(x, mode) : cannot coerce type 'environment' to vector of type 'any'`
     <details>
    <summary>Example of the difference</summary>

    ```r
    > sw_table <- arrow_table(starwars)

    > sw_table |>
    +   filter(height > 100) |>
    +   group_by(homeworld) |>
    +   pull(name) |> 
    +   class()
    [1] "Table"        "ArrowTabular" "ArrowObject"  "R6"          

    > sw_table |>
    +   filter(height > 100) |>
    +   pull(name) |> 
    +   class()
    [1] "ChunkedArray" "ArrowDatum"   "ArrowObject"  "R6"
    ```
    </details>

### CODE REVIEW

**Comment by eitsupi** [comment]
> 2\. Pulling an ungrouped Table produces a ChunkedArray whereas pulling a grouped Table produces a Table. This makes a subsequent call to `as.vector` produce an error of `Error in as.vector(x, mode) : cannot coerce type 'environment' to vector of type 'any'`

This is due to ARROW-17738.
You can use `as_arrow_table` instead of `compute` here I think. https://github.com/apache/arrow/pull/14160#issuecomment-1268404022

**Comment by amoeba** [comment]
Thanks for the link @eitsupi that change fixes the issue I was having.

Re:  @wjones127's comment:

> On the CHECK issue, IIRC we don't @export dplyr S3 methods. We do call s3_register on them here:

I removed the export directive but I would eventually like to understand why the other exported functions in `arrow-tabular.R` (and probably elsewhere) don't need specific documentation (but I'll do that elsewhere).

Both the issues I raised are resolved and this can be reviewed.

File: r/R/arrow-tabular.R [hunk 1]
```diff
@@ -259,3 +259,7 @@ na.omit.ArrowTabular <- function(object, ...) {
 
 #' @export
 na.exclude.ArrowTabular <- na.omit.ArrowTabular
+
+pull.ArrowTabular <- function(x, var = -1) {
```
Comment (nealrichardson) [inline r/R/arrow-tabular.R:263]: I'd move this to dplyr-collect.R to keep it with the other pull() method definitions

**Review by nealrichardson — COMMENTED** [review]
One suggestion but otherwise LGTM, thanks!

**Comment by eitsupi** [comment]
As I mentioned in JIRA, do you implement an argument like `as_data_frame`?
For consistency with other functions, wouldn't it be better to return an R vector by default and an Arrow object optionally?

**Comment by eitsupi** [comment]
Since the dplyr documentation says that pull returns a R vector, I think the documentation should at least explain that `pull` to arrow objects does not return a vector and using with `as.vector` to become a vector.
Users likely know very little about Chuncked Arrays.

File: r/R/arrow-tabular.R (hunk 1 above)
Comment (amoeba) [reply r/R/arrow-tabular.R]: Sounds good. Done in https://github.com/apache/arrow/pull/14330/commits/9c8691bb0f430c1e4a0a62003547fa4e57d62d70.

**Comment by amoeba** [comment]
Thanks the look @nealrichardson. This should be good to merge now.

**Comment by nealrichardson** [comment]
> Since the dplyr documentation says that pull returns a R vector, I think the documentation should at least explain that `pull` to arrow objects does not return a vector and using with `as.vector` to become a vector. Users likely know very little about Chuncked Arrays.

@eitsupi Good call, I'll add a note about this in #14387

**Review by nealrichardson — APPROVED** [review]

**Pull request merged**